### PR TITLE
test: stress test lb with netperf and super_netperf

### PR DESCRIPTION
**Pull request by borkmann**
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6185)

### CODE REVIEW

**Comment by borkmann** [comment]
test-me-please

File: test/runtime/lb.go [hunk 1]
```diff
@@ -274,6 +274,126 @@ var _ = Describe("RuntimeLB", func() {
 			}
 		}, 500)
 
+		It("validates that services work for L4 (IP+Port) loadbalancing, netperf single", func() {
+			err := createLBDevice(vm)
+			if err != nil {
+				log.Errorf("error creating interface: %s", err)
+			}
+			Expect(err).Should(BeNil())
```
Comment (eloycoto) [inline test/runtime/lb.go:282]: `Expect(err).Should(BeNil(), "cannot create LB device")`

File: test/runtime/lb.go [hunk 2]
```diff
@@ -274,6 +274,126 @@ var _ = Describe("RuntimeLB", func() {
...
+				status.ExpectSuccess(fmt.Sprintf("failed run netperf %s against %s", helpers.UDP_RR, ip))
+			}
+		}, 500)
+
+		It("validates that services work for L4 (IP+Port) loadbalancing, netperf stress", func() {
+			err := createLBDevice(vm)
+			if err != nil {
+				log.Errorf("error creating interface: %s", err)
+			}
+			Expect(err).Should(BeNil())
```
Comment (eloycoto) [inline test/runtime/lb.go:342]: `Expect(err).Should(BeNil(), "cannot create LB device")`

File: test/runtime/lb.go [hunk 3]
```diff
@@ -274,6 +274,126 @@ var _ = Describe("RuntimeLB", func() {
...
 
+		It("validates that services work for L4 (IP+Port) loadbalancing, netperf single", func() {
+			err := createLBDevice(vm)
+			if err != nil {
+				log.Errorf("error creating interface: %s", err)
+			}
+			Expect(err).Should(BeNil())
+
+			httpd1, err := vm.ContainerInspectNet(helpers.Httpd1)
+			Expect(err).Should(BeNil())
```
Comment (eloycoto) [inline test/runtime/lb.go:285]: Expect(err).Should(BeNil(), "Cannot get httpd1 IP")

File: test/runtime/lb.go [hunk 4]
```diff
@@ -274,6 +274,126 @@ var _ = Describe("RuntimeLB", func() {
...
+			httpd2, err := vm.ContainerInspectNet(helpers.Httpd2)
+			Expect(err).Should(BeNil())
```
Comment (eloycoto) [inline test/runtime/lb.go:287]: Expect(err).Should(BeNil(), "Cannot get httpd2 IP")

File: test/runtime/lb.go [hunk 5]
```diff
@@ -274,6 +274,126 @@ var _ = Describe("RuntimeLB", func() {
...
+				status := vm.ServiceAdd(svc, fe, be)
+				status.ExpectSuccess("failed to create service %s=>%v", fe, be)
+				svc++
+			}
+
+			By("Making netperf TCP STREAM requests from container => bpf_lb => container")
+
+			for ip := range services {
+				status := vm.ContainerExec(helpers.Client, helpers.Netperf(ip, helpers.TCP_STREAM))
+				status.ExpectSuccess(fmt.Sprintf("failed run netperf %s against %s", helpers.TCP_STREAM, ip))
```
Comment (eloycoto) [inline test/runtime/lb.go:312]: tatus.ExpectSuccess("failed run netperf %s against %s", helpers.TCP_STREAM, ip)

File: test/runtime/lb.go [hunk 6]
```diff
@@ -274,6 +274,126 @@ var _ = Describe("RuntimeLB", func() {
...
+			}
+
+			By("Making netperf TCP RR requests from container => bpf_lb => container")
+
+			for ip := range services {
+				status := vm.ContainerExec(helpers.Client, helpers.Netperf(ip, helpers.TCP_RR))
+				status.ExpectSuccess(fmt.Sprintf("failed run netperf %s against %s", helpers.TCP_RR, ip))
```
Comment (eloycoto) [inline test/runtime/lb.go:319]: fmt.Sprintf is not needed

File: test/runtime/lb.go [hunk 7]
```diff
@@ -274,6 +274,126 @@ var _ = Describe("RuntimeLB", func() {
...
+			}
+
+			By("Making netperf UDP STREAM requests from container => bpf_lb => container")
+
+			for ip := range services {
+				status := vm.ContainerExec(helpers.Client, helpers.Netperf(ip, helpers.UDP_STREAM))
+				status.ExpectSuccess(fmt.Sprintf("failed run netperf %s against %s", helpers.UDP_STREAM, ip))
```
Comment (eloycoto) [inline test/runtime/lb.go:326]: fmt.Sprintf is not needed.

File: test/runtime/lb.go [hunk 8]
```diff
@@ -274,6 +274,126 @@ var _ = Describe("RuntimeLB", func() {
...
+			}
+
+			By("Making netperf UDP RR requests from container => bpf_lb => container")
+
+			for ip := range services {
+				status := vm.ContainerExec(helpers.Client, helpers.Netperf(ip, helpers.UDP_RR))
+				status.ExpectSuccess(fmt.Sprintf("failed run netperf %s against %s", helpers.UDP_RR, ip))
```
Comment (eloycoto) [inline test/runtime/lb.go:333]: fmt.Sprintf is not needed.

File: test/runtime/lb.go [hunk 9]
```diff
@@ -274,6 +274,126 @@ var _ = Describe("RuntimeLB", func() {
...
+
+		It("validates that services work for L4 (IP+Port) loadbalancing, netperf stress", func() {
+			err := createLBDevice(vm)
+			if err != nil {
+				log.Errorf("error creating interface: %s", err)
+			}
+			Expect(err).Should(BeNil())
+
+			httpd1, err := vm.ContainerInspectNet(helpers.Httpd1)
+			Expect(err).Should(BeNil())
```
Comment (eloycoto) [inline test/runtime/lb.go:345]: Expect(err).Should(BeNil(), "Cannot get httpd1 IP")

File: test/runtime/lb.go [hunk 10]
```diff
@@ -274,6 +274,126 @@ var _ = Describe("RuntimeLB", func() {
...
+			httpd2, err := vm.ContainerInspectNet(helpers.Httpd2)
+			Expect(err).Should(BeNil())
```
Comment (eloycoto) [inline test/runtime/lb.go:347]: Expect(err).Should(BeNil(), "Cannot get httpd2 IP")

File: test/runtime/lb.go [hunk 11]
```diff
@@ -274,6 +274,126 @@ var _ = Describe("RuntimeLB", func() {
...
+				status := vm.ServiceAdd(svc, fe, be)
+				status.ExpectSuccess("failed to create service %s=>%v", fe, be)
+				svc++
+			}
+
+			By("Making super_netperf TCP STREAM requests from container => bpf_lb => container")
+
+			for ip := range services {
+				status := vm.ContainerExec(helpers.Client, helpers.SuperNetperf(ip, helpers.TCP_STREAM))
+				status.ExpectSuccess(fmt.Sprintf("failed run super_netperf %s against %s", helpers.TCP_STREAM, ip))
```
Comment (eloycoto) [inline test/runtime/lb.go:372]: fmt.Sprintf is not needed.

File: test/runtime/lb.go [hunk 12]
```diff
@@ -274,6 +274,126 @@ var _ = Describe("RuntimeLB", func() {
...
+			}
+
+			By("Making super_netperf TCP RR requests from container => bpf_lb => container")
+
+			for ip := range services {
+				status := vm.ContainerExec(helpers.Client, helpers.SuperNetperf(ip, helpers.TCP_RR))
+				status.ExpectSuccess(fmt.Sprintf("failed run super_netperf %s against %s", helpers.TCP_RR, ip))
```
Comment (eloycoto) [inline test/runtime/lb.go:379]: fmt.Sprintf is not needed.

File: test/runtime/lb.go [hunk 13]
```diff
@@ -274,6 +274,126 @@ var _ = Describe("RuntimeLB", func() {
...
+			}
+
+			By("Making super_netperf UDP STREAM requests from container => bpf_lb => container")
+
+			for ip := range services {
+				status := vm.ContainerExec(helpers.Client, helpers.SuperNetperf(ip, helpers.UDP_STREAM))
+				status.ExpectSuccess(fmt.Sprintf("failed run super_netperf %s against %s", helpers.UDP_STREAM, ip))
```
Comment (eloycoto) [inline test/runtime/lb.go:386]: fmt.Sprintf is not needed.

File: test/runtime/lb.go [hunk 14]
```diff
@@ -274,6 +274,126 @@ var _ = Describe("RuntimeLB", func() {
...
+			}
+
+			By("Making super_netperf UDP RR requests from container => bpf_lb => container")
+
+			for ip := range services {
+				status := vm.ContainerExec(helpers.Client, helpers.SuperNetperf(ip, helpers.UDP_RR))
+				status.ExpectSuccess(fmt.Sprintf("failed run super_netperf %s against %s", helpers.UDP_RR, ip))
```
Comment (eloycoto) [inline test/runtime/lb.go:393]: fmt.Sprintf is not needed.

**Review by eloycoto — CHANGES_REQUESTED** [review]
Love the changes, a few minor things.

File: test/helpers/wrappers.go [hunk 15]
```diff
@@ -77,6 +75,12 @@ func Netperf(endpoint string, perfTest PerfTest) string {
 	return fmt.Sprintf("netperf -l 3 -t %s -H %s", perfTest, endpoint)
 }
 
+// SuperNetperf returns the string representing the super_netperf command to use
+// when testing connectivity between endpoints.
+func SuperNetperf(endpoint string, perfTest PerfTest) string {
+	return fmt.Sprintf("super_netperf 200 -l 30 -t %s -H %s", perfTest, endpoint)
```
Comment (jrfastab) [inline test/helpers/wrappers.go:81]: I have a version of this that takes the number of netperf instances as input. Can we make this
`func SuperNetperf(endpoint string, perfTest PerfTest, instances int) string {...}` Or I can just make it another routine.

**Pull request closed without merging**